### DISCUSSION TITLE
Why is PyObject not JSON Serializable?

### DISCUSSION THREAD

**Discussion by maxwelllevin** [pydantic/pydantic#3842] [Question]
Hi 👋🏻 

I'm working on a project where we plan to use pydantic to validate configuration files. One of the things we want to support is dependency injection via a parameter, which is a dotted-path to an importable python class or function (e.g., `'math.cos'`).

I think the pydantic `PyObject` type supports this quite nicely, except that it is not currently JSON serializable. This is unfortunate, as it means the `PyObject` type can't be used in models that will be turned into JSON or, in my case, used to generate JSON schema to help validate configuration files.

I've searched through all the pydantic documentation, issues, and discussions involving `PyObject`, and the only thing I've found that's relevant to my question is #364, in which there was a brief discussion about JSON schema support for callable types. The result of that conversation was that no, JSON schema does not support callable types so any pydantic fields that are callables should be excluded from JSON schema. Part of that discussion included [this note from @samuelcolvin](https://github.com/samuelcolvin/pydantic/pull/364#issuecomment-456476606):
> `PyObject` is a bit different as it is passed as just a string. It would obviously be very dangerous to do if you don't have control over who can submit data, but in theory, there's nothing to stop someone submitting `'math.cos'` or any other valid path to a `PyObject` field.

which makes a distinction between `PyObject` and other `Callable` types, and seems to me to be in favor of adding support for PyObject in JSON schema via a string (e.g., `'math.cos'` instead of `<built-in function cos>`), though that could be wishful thinking on my part.

@tiangolo's response to this seems to be focused on `typing.Callable` (that *was* the main point of the PR, so totally fair):
> The closest I see would be a string with a `format` of some type (as an extension, which is valid, that's the case for `timedelta`), but I don't see a way to put it that makes sense for a Python callable, so I still think option 2 would be the most reasonable.

Here option 2 is to exclude callable fields from the schema and include a warning. I think this makes sense for callables in general, but I feel like the discussion on `PyObject` specifically wasn't really resolved. Aside from these two comments I haven't been able to find any discussions on `PyObject` in the JSON schema.

I am wondering:
- Are the authors open to making `PyObject` serializable as its input string?
- Would people other than myself benefit from this?
- What challenges or roadblocks would there be in implementing something like this?

**Comment by tiangolo** [comment]
I think the first task would probably be to find what use case this would solve, and check if it can't be solved in another way already.

The next would be to think, what would you imagine would go in JSON Schema for an arbitrary Python object using `PyObject`? And what use would it have in JSON Schema?
